state/upgrades_test.go

```diff
@@ -207,3 +214,120 @@ func (s *upgradesSuite) TestSplitMigrationStatusMessages(c *gc.C) {
+	state1 := s.makeModel(c, "m1", coretesting.Attrs{}, ModelArgs{Type: ModelTypeCAAS})
+	state2 := s.makeModel(c, "m2", coretesting.Attrs{}, ModelArgs{Type: ModelTypeCAAS})
+	defer func() {
+		_ = state1.Close()
+		_ = state2.Close()
+	}()
+
+	appColl1, closer := state1.db().GetRawCollection(applicationsC)
+	defer closer()
+
+	model1, err := state1.Model()
+	c.Assert(err, gc.IsNil)
+
+	err = appColl1.Insert(bson.M{
+		"_id":        ensureModelUUID(model1.UUID(), "app1"),
+		"name":       "app1",
+		"model-uuid": model1.UUID(),
+	})
+	c.Assert(err, gc.IsNil)
+	err = appColl1.Insert(bson.M{
+		"_id":        ensureModelUUID(model1.UUID(), "app2"),
+		"name":       "app2",
+		"model-uuid": model1.UUID(),
+	})
+	c.Assert(err, gc.IsNil)
+	// app3 does not get backfilled because its storage unique ID is already
+	// populated.
+	err = appColl1.Insert(bson.M{
+		"_id":               ensureModelUUID(model1.UUID(), "app3"),
+		"name":              "app3",
+		"model-uuid":        model1.UUID(),
+		"storage-unique-id": "uniqueid3",
+	})
+	c.Assert(err, gc.IsNil)
+
+	model2, err := state2.Model()
+	c.Assert(err, gc.IsNil)
+
+	appColl2, closer := state2.db().GetRawCollection(applicationsC)
+	defer closer()
+
+	err = appColl2.Insert(bson.M{
+		"_id":        ensureModelUUID(model2.UUID(), "app4"),
+		"name":       "app4",
+		"model-uuid": model2.UUID(),
+	})
+	c.Assert(err, gc.IsNil)
+	err = appColl2.Insert(bson.M{
+		"_id":        ensureModelUUID(model2.UUID(), "app5"),
+		"name":       "app5",
+		"model-uuid": model2.UUID(),
+	})
+	c.Assert(err, gc.IsNil)
+
+	appMigratedCount := 0
+
+	getStorageUniqueID := func(
+		apps []AppAndStorageID,
+		model *Model,
+	) ([]AppAndStorageID, error) {
+		fakeK8s := map[string][]appNameAndID{
+			model1.UUID(): {
+				{
+					appName:  "app1",
+					uniqueID: "uniqueid1",
+				},
+				{
+					appName:  "app2",
+					uniqueID: "uniqueid2",
+				},
+				{
+					appName:  "app3",
+					uniqueID: "uniqueid3",
+				},
+			},
+			model2.UUID(): {
+				{
+					appName:  "app4",
+					uniqueID: "uniqueid4",
+				},
+				{
+					appName:  "app5",
+					uniqueID: "uniqueid5",
+				},
+			},
+		}
+
+		appsAndStorageIDs := make([]AppAndStorageID, 0, len(apps))
+		k8sDeployment, ok := fakeK8s[model.UUID()]
+		if !ok {
+			return nil, errors.Errorf("unknown model %q", model.UUID())
+		}
+		for _, app := range apps {
+			index := slices.IndexFunc(k8sDeployment, func(a appNameAndID) bool {
+				return a.appName == app.Name
+			})
+			if index == -1 {
+				c.Fatalf("app %q not found, this should not happen", app.Name)
+			}
+
+			appMigratedCount++
+			appsAndStorageIDs = append(appsAndStorageIDs, AppAndStorageID{
+				Id:              app.Id,
+				Name:            app.Name,
+				StorageUniqueID: k8sDeployment[index].uniqueID,
+			})
+		}
+
+		return appsAndStorageIDs, nil
+	}
+
+	err = PopulateApplicationStorageUniqueID(s.pool, getStorageUniqueID)
+	c.Assert(err, gc.IsNil)
+	c.Assert(appMigratedCount, gc.Equals, 4)
```

Comment:
We're not asserting the value of content of the storage id written to the database that I can see. We need to check any existing value (for app3) is not backfilled and the new ids get written as expected.